x86 EFI improvements

tbursztyka opened the pull request:
- Exposing EFI into Zephyr: so other parts - such as ACPI - can take advantage of it
- To mitigate whether it booted through multiboot or EFI, adding a boot argument


Side notes:
I do not pass the EFI image handle to zephyr, so it is not possible to call ExitBootServices(). Maybe we could in the future. (So EFI subsystem will let go its console hooks and all the boot services).

Commit:
arch/x86: Installing generic EFI header

This will prove to be useful for ACPI later, among other possible
things.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
zefi: Use generic EFI header and remove local one

Apply the changes so zefi looks for zephyr includes.

up_squared board being the only one using zefi, update its
CMakeLists.txt so it adds the relevant include dir.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: Expose EFI configuration lookup function

This will be useful in ACPI, at least.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: Lookup for ACPI RSDP through EFI

Now that we can look up into EFI's configuration table, let's do so to
find ACPI's RSDP.

Both ACPI version 1 and 2 (and later) are looked up.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: Initialize EFI system table pointer for ia64

Pass the pointer to ebx as for multiboot structure via zefi,
and let's use this properly in prep_c.c

Then, it is possible to store this pointer and have access to the EFI
system table at any time after booting up.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
boards: Enable EFI system table support on up_squared

This is meant to work for the zefi built image.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: Expose multiboot init function even when disabled

Just a dummy function will do.

When enabled, the code does not need the #ifdef as cmake is handling
this properly already. This was also the wrong CONFIG_ used there
anyway.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
arch/x86: Expand cpu boot argument

In order to mitigate at runtime whether it booted on multiboot or EFI,
let's introduce a dedicated x86 cpu argument structure which holds the
type and the actual pointer delivered by the method (multiboot_info, or
efi_system_table)

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
I have issues atm, on hello world sample:
- on x86_64 + EFI: I get only "He" as a message. used to work before, so I probably did a mistake in rebasing. (it works perfectly fine on x86_64 + multiboot)
- on x86 + multiboot: it does not seem to work, probably my assembly is messed up. 

I'll check.

andrewboie commented:
Code and overall structure LGTM. I can't spot yet what might be causing the hello_world issues you are seeing.

andyross commented on include/arch/x86/multiboot.h:
```diff
@@ -36,7 +36,18 @@ struct multiboot_info {
-extern void z_multiboot_init(struct multiboot_info *);
+#ifdef CONFIG_MULTIBOOT_INFO
+
+void z_multiboot_init(struct multiboot_info *info);
```
Since this is doing multiple duty now as both "multiboot" and "our own boot hook", maybe we should rename it?

tbursztyka replied:
Or I could test the arg pointer that is supposed to hold multiboot_info or efi_sys_table: if NULL then it's neither multiboot nor efi.

jenmwms replied:
Perhaps this is a bit confusing? I don't understand why we have this #ifdef CONFIG_MULTIBOOT_INFO here, it seems to be expected - line 37 defines the struct multiboot_info, which to me indicates we have it configured. I'm sure I'm missing something, sorry!

tbursztyka replied:
It's just to avoid using #ifdef extensively in the .c files, I prefer using if (IS_ENABLED()) { ... }, so in order to do it I need to enclose the exposed functions with #ifdef in the .h files. (the structures, and macros don't need to)

andyross commented on arch/x86/core/ia32/crt0.S:
```diff
@@ -260,7 +261,13 @@ __csSet:
 
-	pushl	%ebx		/* pointer to multiboot info, or NULL */
+	movl	$x86_cpu_boot_arg, %ebp
+	/* Boot type to multiboot, ebx content will help to mitigate */
```
Not sure there's much value in this being a runtime detection.  It doesn't hurt anything really, but in general Zephyr always knows the platform is running on and we'll never need to boot the same binary from both grub and EFI.

tbursztyka replied:
I did not change the existing build, i.e.: both zephyr.strip and zephyr.efi are built. Perhaps that would be the next step, when zefi will be using the same toolchain and will be aware of the CONFIG_*. At this point it will be possible indeed to get rid of the boot_type (unless we make this zefi work first, then I'll rebase this PR against it)

andyross reviewed (commented):
Handful of notes.   This all looks really good to me.

jenmwms commented on arch/x86/core/intel64/locore.S:
```diff
@@ -157,8 +157,13 @@ __start:
+	/* and multiboot info */
+	movl %ebx, (__x86_cpuboot_t_arg_OFFSET + \
+		    __x86_boot_arg_t_arg_OFFSET)(%ebp)
 
```
To me, this looks to be inserting boot type and multiboot info at the same location. They have the same offsets from %edp, or am I misunderstanding this?

jenmwms replied:
oops, I see the difference now. Sorry!

jenmwms reviewed (approved):
> * on x86_64 + EFI: I get only "He" as a message. used to work before, so I probably did a mistake in rebasing. (it works perfectly fine on x86_64 + multiboot)
> * on x86 + multiboot: it does not seem to work, probably my assembly is messed up.
> 
> I'll check.

The changes LGTM but please let us know what you find when you recheck ^^

jhedberg reviewed (changes requested):
Maybe I'm doing something wrong, but I took this PR, rebased it against latest master, and tried to build `samples/subsys/shell/shell_module` for `up_squared` with the following result:
```
[139/139] Linking C executable zephyr/zephyr.elf
FAILED: zephyr/zephyr.elf zephyr/zephyr.bin zephyr/zephyr.lst zephyr/zephyr.stat zephyr/zephyr.strip 
: && ccache /opt/toolchains/zephyr-sdk-0.11.2/x86_64-zephyr-elf/bin/x86_64-zephyr-elf-gcc -m64   zephyr/CMakeFiles/zephyr_final.dir/misc/empty_file.c.obj  -o zephyr/zephyr.elf  zephyr/arch/arch/x86/pagetables.o  -Wl,-T  zephyr/linker_pass_final.cmd  -Wl,-Map=/workdir/zephyr/build/zephyr/zephyr_final.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/arch/x86/core/libarch__x86__core.a  zephyr/lib/libc/minimal/liblib__libc__minimal.a  zephyr/lib/posix/liblib__posix.a  zephyr/soc/x86/apollo_lake/libsoc__x86__apollo_lake.a  zephyr/drivers/serial/libdrivers__serial.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  zephyr/CMakeFiles/offsets.dir/arch/x86/core/offsets/offsets.c.obj  -L"/opt/toolchains/zephyr-sdk-0.11.2/x86_64-zephyr-elf/bin/../lib/gcc/x86_64-zephyr-elf/9.2.0"  -L/workdir/zephyr/build/zephyr  -lgcc  -m64  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -no-pie  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn && cd /workdir/zephyr/build/zephyr && /usr/local/bin/cmake -E rename zephyr_final.map zephyr.map && /opt/toolchains/zephyr-sdk-0.11.2/x86_64-zephyr-elf/bin/x86_64-zephyr-elf-objcopy --gap-fill 0xff --output-target=binary --remove-section=.comment --remove-section=COMMON --remove-section=.eh_frame zephyr.elf zephyr.bin && /opt/toolchains/zephyr-sdk-0.11.2/x86_64-zephyr-elf/bin/x86_64-zephyr-elf-objdump -d -S zephyr.elf >zephyr.lst && /opt/toolchains/zephyr-sdk-0.11.2/x86_64-zephyr-elf/bin/x86_64-zephyr-elf-readelf -e zephyr.elf > zephyr.stat && /opt/toolchains/zephyr-sdk-0.11.2/x86_64-zephyr-elf/bin/x86_64-zephyr-elf-strip --strip-all zephyr.elf -ozephyr.strip && /usr/bin/python3.6 /workdir/zephyr/arch/x86/zefi/zefi.py -i /workdir/zephyr/include -f /workdir/zephyr/build/zephyr/zephyr.elf
In file included from /workdir/zephyr/arch/x86/zefi/zefi.c:7:0:
/workdir/zephyr/include/arch/x86/efi.h: In function ‘efi_config_get_vendor_table_by_guid’:
/workdir/zephyr/include/arch/x86/efi.h:599:2: error: implicit declaration of function ‘ARG_UNUSED’ [-Werror=implicit-function-declaration]
  ARG_UNUSED(guid);
  ^~~~~~~~~~
/workdir/zephyr/include/arch/x86/efi.h:601:9: error: ‘NULL’ undeclared (first use in this function)
  return NULL;
         ^~~~
/workdir/zephyr/include/arch/x86/efi.h:601:9: note: each undeclared identifier is reported only once for each function it appears in
cc1: all warnings being treated as errors
Traceback (most recent call last):
  File "/workdir/zephyr/arch/x86/zefi/zefi.py", line 160, in <module>
    build_elf(args.elf_file, args.include)
  File "/workdir/zephyr/arch/x86/zefi/zefi.py", line 119, in build_elf
    subprocess.run(cmd, check = True)
  File "/usr/lib/python3.6/subprocess.py", line 438, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command '['gcc', '-shared', '-Wall', '-Werror', '-I.', '-I', '/workdir/zephyr/include', '-fno-stack-protector', '-fpic', '-mno-red-zone', '-fshort-wchar', '-Wl,-nostdlib', '-T', '/workdir/zephyr/arch/x86/zefi/efi.ld', '-o', 'zefi.elf', '/workdir/zephyr/arch/x86/zefi/zefi.c']' returned non-zero exit status 1.
ninja: build stopped: subcommand failed.
FATAL ERROR: command exited with status 1: /usr/local/bin/cmake --build /workdir/zephyr/build
XIO:  fatal IO error 4 (Interrupted system call) on X server ":0"
      after 863 requests (863 known processed) with 0 events remaining.
```

Seems like the problem is `efi.h` missing e.g. `#include <stddef.h>` as well as `#include <autoconf.h>`, however the latter has the additional problem that the generated include directory is not made available to the include lookup paths of the `zefi.c` build

tbursztyka commented:
@jhedberg looks like I merged a modification that was not ready to get in.

jhedberg commented:
> @jhedberg looks like I merged a modification that was not ready to get in.

Ok, looking forward to a fixed version. That said, it seems we're missing some CI hooks, since stuff like this should ideally be caught by that.

jhedberg commented:
@tbursztyka what's the plan/schedule for getting this one merged?

Commit:
efi: Passing the image handle to zephyr along with the efi system table

zefi now pushes the image handle as well, and x86_64 arch takes provides
it to efi subsystem.

This will be useful for calling efi services that require such handle.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
efi: Add a function to exit boot services

This will shut down all EFI boot services so Zephyr becomes the only one
in charge of the system.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

jenmwms commented:
Nice work! Please update the description to reflect some recent commits (esp `35551b5` "Passing the image handle to zephyr along with the efi system table")

tbursztyka commented:
> I have issues atm, on hello world sample:
>
>    on x86_64 + EFI: I get only "He" as a message. used to work before, so I probably did a mistake in rebasing. (it works perfectly fine on x86_64 + multiboot)
    on x86 + multiboot: it does not seem to work, probably my assembly is messed up.

This comment I made last october is still valid. I had no time to work on this since.

Commit:
efi: Adding status code to header

This will be usefull when calling EFI functions.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
Well, this PR is pointless. I made the dumb assumption that EFI memory space could be mapped within Zephyr, and worse that EFI boot services could be called within Zephyr...
Which is not. If I had read UEFI doc more carefully I would have known that this can only be manipulated in the EFI stub (thus inside zefi).

Closing then. Dunno if I'll get time to move the logic into zefi someday...
